### ISSUE TITLE
Markdown parsing of descriptions

### ISSUE THREAD

**Issue by nodiscc** [shaarli/Shaarli#206]
Add an option to render link descriptions from markdown to HTML.
Should be a plugin, blocked by https://github.com/shaarli/shaarli/issues/14, https://github.com/shaarli/Shaarli/pull/164

**Comment by alexisju** [comment]
I have tested the implementation of parsedown from "shinterest" https://github.com/broncowdd/Shinterest. It's basically working but still buggy (ie link parsing [text](http://link) doesn't work).

--
I have also tried the zepto.js method. It is supposed working but not easy to use for the user:
http://user23.net/bin/?3fba27786063f200#JdFe9TdWx+2TVMFiFLtTqYAQfpgoXeZzIbn8W75blC4=

--

I've also tried js client methods but it slow down shaarli to much
https://github.com/p01/mmd.js
https://github.com/dnordstrom/parseMarkdown.js
https://github.com/borgar/textile-js

https://github.com/dtjm/plaintext.js

--
Finaly, for shaarli, php method remain the best. I've tried to enable it into a template but it seems difficult (due to rainTPL). It need to be implemeted into core shaarli( maybe not as module).

Parsedown seems to be a good way:
https://github.com/erusev/parsedown

**Comment by alexisju** [comment]
Maybe is it possible to to adapt Timo's code from BlogoText ?

https://github.com/timovn/blogotext/blob/master/inc/conv.php

**Comment by alexisju** [comment]
Implemented into this fork : https://github.com/Knah-Tsaeb/Shaarli/commit/6f4fd910a9fd7786a7b135b0dc5b2a77d0b1028f

**Comment by alexisju** [comment]
@nodiscc you tagged this issue as plugin : is it possible to load external includes (ie: parsedown.php) into template? (when i tried i got errors from RainTPL)

**Comment by nodiscc** [comment]
>is it possible to load external includes (ie: parsedown.php)

Ideally it should be, but the plugin system is unfinished work. A markdown plugin would look like
```
└── tpl
    └── plugins
        └── markdown
            ├── README.md
            ├── markdown.description.html
            ├── markdown.png
            └── php-markdown.php
```

where all rendering of the description area in the linklist is left to the markdown plugin.

https://github.com/piero-la-lune/Bumpy-Booby uses PHP-Markdown and it works wonderfully. <sup>by the way this is a very good bug/issue tracker and you should try it if you need a personal, self-hosted issue tracker</sup>

**Comment by webdev23** [comment]
Hi everyone. I am the author of the "zepto.js" that alexisju linked up: 
Retrouver (la syntaxe perdue de) Shaarli:
http://user23.net/links/?08KERw
http://user23.net/bin/?3fba27786063f200#JdFe9TdWx+2TVMFiFLtTqYAQfpgoXeZzIbn8W75blC4=

I just tested this on a fresh install and it work fluently. Here is a more accurate description. We can also use jquery, already included (but 3 times bigger):
So by adding, in /tpl/includes.html, line 10:
<script src="../inc/jquery-1.11.2.min.js"></script>

In /tpl/linklist.html:
On line 45, adding a variable ID (replace all the line by this one):
{if="$value.description"}<div id="ldesc_{$value.linkdate}" class="linkdescription">{$value.description}</div>{/if}

On line 66, this script block:
<script>
		 $(function() {
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/{pic <a href="/g,'<img style="display:block;max-width:50%;margin:15px 0 0 5px;height:auto" src="'));
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/{pic. <a href="/g,'<img style="display:block;max-width:50%;float:right;margin:15px 0 0 5px;height:auto" src="'));		   
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/pic}/g,'<br>'));
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/{3/g,'<h3>')); 
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/3}/g,'</h3>'));
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/{2/g,'<h2>')); 
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/2}/g,'</h2>'));
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/{1/g,'<h1>')); 
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/1}/g,'</h1>'));
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/{6/g,'<h6>')); 
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/6}/g,'</h6>'));
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/{5/g,'<h5>')); 
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/5}/g,'</h5>'));
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/{4/g,'<h4>')); 
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/4}/g,'</h4>'));
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/{i/g,'<span style="font-style:italic">')); 
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/i}/g,'</span>'));
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/{lt/g,'<span style="text-decoration:line-through">')); 
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/lt}/g,'</span>'));
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/{s/g,'<span style="font-weight: bold;font-size:1.15em">')); 
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/s}/g,'</span>')); 
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/{m/g,'<small>')); 
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/m}/g,'</small>'));                       
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/{audio <a href="/g,'<audio style="margin:6 0 3 2" controls><source src="'));
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/{video <a href="/g,'<video type="video/mp4" style="margin:6 10 3 2;max-width:66%;height:auto" controls><source src="'));
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/video}/g,'</video>'));
	       $('#ldesc_{$value.linkdate}').html($('#ldesc_{$value.linkdate}').html().replace(/audio}/g,'</audio>'));       		   		   
           });
		</script>

Work directly, like this:*
{pic http://mypic.com pic}
{audio http://mypic.com audio}
{1 My big tittle h1 1}
{lt line-trought barré lt}
(..)

I am trying to adapt this as a plugin. I get stucked on something. The loop is not working when i put the script part in the plugin folder. 
└── tpl
    └── plugins
        └── shaardown
            ├── shaardown.linklist.html

And declare it as a plugin on index.php line 36:
$GLOBALS['config']['PLUGINS'] = array('archiveorg', 'playvideos', 'qrcode', 'readityourself', 'wallabag','shaardown'); // List of enabled plugins

I think it should work like this out of the box?
What i missed for make the loop work?
What do you think of the syntax? The real markdown (with a lot of ###) is not simple to filter, that's why i used the {} at this time.
Plugin system is a nice idea but i don't get it now!
Cheers brothers!

**Comment by nodiscc** [comment]
Hi @webdev23 the plugin system is currently unfinished and you'd have to extend it before being able to parse descriptions with it. See the [commit messages in the `new-plugin-sytem` branch](https://github.com/nodiscc/Shaarli/commits/new-plugin-system) to get an idea of how it works/is extensible. Someone may be able to help us (I won't have time to finish it soon), but you should really put up a git repository showing the exact changes you applied.

I should have said _A markdown plugin would look like [...] when parsing descriptions through a plugin is implemented_. The plugin system currently only allows adding custom HTML code to `<head>`, toolbar section and link buttons (date/qr code area). Sorry if I was unclear.


Other than that, I think the markdown syntax is already widely used, known, feature complete and already existing markdown parsers are well tested and spare us most of the work (writing a custom syntax) - that's why I'm interested in a markdown plugin. But a custom plugin that parses descriptions in whatever other markup could be a first step, so thanks for looking into this.

**Comment by alexisju** [comment]
Should we really "wait" a working plugin system rather implement md by default?

Again, the code from Knah-Tsaeb seems clean and could be pushed* into current shaarli. It seems standardised md and reliable (from parsedown.org). 

*We just need to adapt the css to be sure all works.

Commit : https://github.com/Knah-Tsaeb/myShaarli/commit/6f4fd910a9fd7786a7b135b0dc5b2a77d0b1028f

You can test md here : http://effingo.be/test/shaarli-md/ (test/test)

**Comment by ArthurHoaro** [comment]
Maybe we should rather go for a client side Markdown renderer (JS).

The main advantage is that we don't need to alter data (existing links) when it is enabled/disabled.

**Comment by alexisju** [comment]
I tested some in js but it slowdown the rendering of shaarli...

**Comment by ArthurHoaro** [comment]
Yeah forget that. We can render it before displaying it, without altering data. Didn't had enough coffee yet. :)

**Comment by nodiscc** [comment]
This should still be off by default;

 * Enabling it for previously saved links would lead to display problems for plain text descriptions that contains characters such as `_*[]` and html tags.
 * When shaaring a link from the bookmarklet, with some text selected, the selected text should not contain any of these characters, otherwise they will be interpreted and rendered as markdown (whereas on the source page, they are just plain text)

**Comment by alexisju** [comment]
I'm not sure to understand what can be the problem. It is very rare to met an article who uses the md syntax by accident. And if it happens, or if _ * or [ ] are in a text, it cannot do something wrong, even using the bookmaklet or do a copy/past.

Moreover, in our case md is supposed only to parse description (not tags) and the parsedown classe  doesn't interpret html.

But maybe it is more convenient to enable (or not) this functionaly by the plugin system who is comming.

**Comment by ArthurHoaro** [comment]
ping #282

**Comment by ArthurHoaro** [comment]
Fixed with plugin in #282.
